Move docs for sealed sender v1/v2 to the appropriate method docstrings

cosmicexplorer opened the pull request:
Hello! I am going to finish up my other pending PRs to this repo soon :).

### Problem
The sealed sender single-recipient and multi-recipient KEM schemes in Sealed Sender v1/v2 are super neat, and documentation for how they work exists in some (very helpful!) source comments. However, it's not immediately clear e.g. that the Signal server receives the result of `sealed_sender_multi_recipient_encrypt()`, splits the output into constituent sealed-sender v2 messages, then routes them individually to clients for `sealed_sender_multi_recipient_decrypt()`. Additionally, the design tradeoffs between v1 and v2 sealed sender mechanisms in general are not made clear in documentation of these public methods.

### Solution
1. Move the (very useful!) documentation for the mechanism of v1 and v2 sealed sender into the docstrings for methods `sealed_sender_encrypt_from_usmc()` and `sealed_sender_multi_recipient_encrypt()`.
2. Clarify the role of `sealed_sender_multi_recipient_fan_out()` in testing.
    - Clarify how the Signal server will split out sealed sender v2 bulk messages and route them to the appropriate recipients on the backend in `sealed_sender_multi_recipient_encrypt()`.
3. Produce a section "Contrast with Sealed Sender v1" in the docstring for `sealed_sender_multi_recipient_encrypt()` which describes the pros and cons of the single-recipient vs multi-recipient KEM schemes implemented in the `sealed_sender_v1` and `sealed_sender_v2` modules.
4. Modify the definition of `sealed_sender_v1::{EphemeralKeys,StaticKeys}` to avoid repeated `Result<Box<...>>`ing, and make `StaticKeys::calculate()` accept `&IdentityKeyPair` and `&IdentityKey` instead of structs from `curve.rs`.
5. Make `sealed_sender_v2::{apply_agreement_xor,compute_authentication_tag}` accept and return fixed-size arrays instead of unsized slices.
6. Avoid separating the AES-GCM-SIV authentication `tag` from the rest of the `ciphertext` in `sealed_sender_multi_recipient_encrypt()` in order to differentiate it from the asymmetric authentication tags `at_i` computed later in the function when looping over the recipients.

### Result
A rendered version of the docs can be seen here: https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/.

This is a great codebase!!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1006,7 +1025,15 @@ pub async fn sealed_sender_multi_recipient_encrypt<R: Rng + CryptoRng>(
 }
 
-// For testing
+/// Split out the result of [sealed_sender_multi_recipient_encrypt] into a sequence of individual
+/// encrypted [UnidentifiedSenderMessageContent]s.
+///
+/// This method is primarily intended for testing in the context of the Signal app, but is also used
+/// e.g. in the jni bindings with
+/// [crate::protocol::SealedSender_MultiRecipientMessageForSingleRecipient] to allow processing a v2
+/// sealed-sender message directly by [sealed_sender_decrypt_to_usmc]. For the Signal app, this
+/// process of splitting out a v2 sealed-sender message into individual messages is typically
+/// performed implicitly by the Signal server.
 pub fn sealed_sender_multi_recipient_fan_out(data: &[u8]) -> Result<Vec<Vec<u8>>> {
```
Ah, *both* uses are for unit testing (not app-level testing) – it's just that one is testing at the Rust level and the other is testing at the Java/Swift/TypeScript level. There's never a real-life need to encrypt something and immediately decrypt it, but that's what the tests do to simulate both sides of a connection. And this isn't the API the server would actually use, because the server needs to keep the association between messages and recipients. So it really is just for testing.

cosmicexplorer replied:
Ah ok thank you! I think I was vaguely approaching this understanding since I posted this--it was at first somewhat confusing to me why this method was tossing away all the info we'd carefully added in `multi_recipient_encrypt()`!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
````diff
@@ -757,37 +757,56 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
+/// mechanism is summarized in pseudocode below.
+///
+/// ### Pseudocode
+///```notrust
````
Nitpick: The rustdoc book suggests `text`, which I like a little better because I initially read this as "no trust" rather than "not rust"!

cosmicexplorer replied:
Oh yes!! I thought that was kind of funny but it's a good idea to be clear about our terminology when we get the chance to be in these matters :)

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -757,37 +757,56 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
+///    [sealed_sender_v2::compute_authentication_tag].
+/// 4. use the keypair from (1) to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent].
+/// 5. As described in ["Randomness Reuse: Extensions and Improvements"], the randomness from (1) is
+///    reused for all of the intended recipients of the message.
```
Looks like this is covered by your note below.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -757,37 +757,56 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// ### Overview
+/// 1. generate a random [sealed_sender_v2::DerivedKeys].
+/// 2. encrypt the keypair using a shared secret derived from the sender's private key and
+///    recipient's public key (this secret never changes for that pair of identity keys) with
+///    [sealed_sender_v2::apply_agreement_xor].
+/// 3. sign the encrypted keypair using the sender's private key with
+///    [sealed_sender_v2::compute_authentication_tag].
+/// 4. use the keypair from (1) to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent].
```
This is a little oversimplified. Maybe it's
1. Generate a series of random bytes.
2. Derive an ephemeral key pair from (1).
3. Encrypt (1) using a shared secret derived from the private ephemeral key (2) and the recipient's public identity key…
4. Add an authentication tag for (3) using a secret derived from the sender's private identity key and the recipient's public identity key.
5. Use the key pair from (2) to…
6. Send the public ephemeral key (2), the encrypted random bytes (3), and the authentication tag (4) to the recipient along with the encrypted message (5).

But also while I think you've done a nice job summarizing what the algorithm does, this is all still implementation detail docs (it's on a private module, after all), and we already have the crypto-paper-style pseudocode (not my personal favorite, but conventional!). What made you decide to add a prose high-level summary of the algorithm? Clearly you felt it was worth it or you wouldn't have bothered to write it out!

cosmicexplorer replied:
So I'm totally fine leaving it the way it is if you'd prefer to keep the private implementation private, as this PR was largely an excuse to hash out my understanding of this process with you! My other reasoning for making this PR was to alleviate my confusion on which parts of this process were handled by the Signal backend server, since in my fork I have extended this sealed-sender v2 mechanism to use when sending encrypted `PreKeyBundle`s (see https://github.com/cosmicexplorer/libsignal-client/commit/a20a58e5aa85d3d6571f4f4e915ff2ef2c5e459a for a very quick summary of the extensions). I was a bit flummoxed at first by what actually went into sealed-sender messages after 18463e8357ef6c8489568b0c0489c2161740842a, since the messages sent over the wire (reasonably!) no longer corresponded directly to a protobuf schema.

Context: I am orchestrating this "encrypted `PreKeyBundle`" mechanism from within an (also AGPL) crate (not public yet) that adapts this codebase to be used in places where the `gpg` tool is normally used, i.e. for sending encrypted messages across insecure channels. For my next trick I would like to also extend the sealed-sender concept to encrypt the message recipient as well, since my impression is that unlike the Signal app which requires the recipient address for routing (mentioned in passing in [this blog post](https://signal.org/blog/sealed-sender/)), my tool will be used in a context where I can ensure messages get routed securely without leaking metadata (even if this requires additional overhead vs the Signal app). If I succeed in doing this, I would actually like to propose pushing these changes upstream, as I see them as enabling additional metadata resistance for Signal conversations, but I wanted to spend the time to develop these techniques first in the context of my own project before doing so.

So to reiterate, the addition of this rather intense docstring is motivated by my own application of the Signal client codebase, which has explicitly different goals than the mainstream Signal app. I accept that upstream will likely change these sorts of implementation details in the future and have been eagerly rebasing against them. I do *not* want to slow down the Signal app by [federating it and making it beholden to lots of external projects that require a stable API](https://signal.org/blog/the-ecosystem-is-moving/). So, feel free to propose closing this PR if you're uncomfortable with that.

jrose-signal reviewed (commented):
Thank you!

jrose-signal commented:
There's a public/private distinction that's important here: `sealed_sender_multi_recipient_encrypt` and `sealed_sender_decrypt` are public, but `sealed_sender_v1` and `sealed_sender_v2` are implementation details. Whether or not the public entry points mention the specific cryptography involved could go either way, but it can't reference private stuff if we're documenting this like a usual crate.

As an aside, there's nothing *wrong* with Sealed Sender v1. In fact, it's a smaller payload on the receiving side than Sealed Sender v2. But its encryption key is derived at least partially from the recipient's identity key, so even if the same content is sent to multiple recipients, it has to be re-encrypted each time. SSv2's real change is separating the encryption key from the recipient's identity key, so that only a per-recipient needs to be encrypted differently.

cosmicexplorer commented:
> SSv2's real change is separating the encryption key from the recipient's identity key

Ok, thanks! That was super helpful! In light of this I have expanded the scope of this PR to also cover documenting the mechanics of v1 sealed sender in `sealed_sender_encrypt_to_usmc()`, as well as a quick comparison of the design tradeoffs of the two mechanisms that you've just summarized!

I also made a quick change just now in a3e0031 to explicitly append the AES-GCM-SIV authentication tag to the end of the symmetric ciphertext in `sealed_sender_multi_recipient_encrypt()`, which I thought made it more clear that *that* authentication tag has nothing to do with the asymmetric authentication tags `at_i` computed later in the method.

All of the docs I've just added also explicitly avoid referring to "v1" or "v2" sealed sender mechanisms (except in passing in `sealed_sender_decrypt_from_usmc()`), instead referring to the single- and multi-recipient KEM schemes described in the docstrings for `sealed_sender_encrypt_to_usmc()` and `sealed_sender_multi_recipient_encrypt()`.

It was a fun exercise writing these docs but if they're more than you'd like to review / aren't appropriate for the repo for some reason feel free to tell me to cut them down!

cosmicexplorer commented:
I've pushed a rendered version of the docs from `cargo doc` to my fork's `gh-pages` branch, viewable at https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/. The two method docstrings which are the subject of this PR:
- [`sealed_sender_encrypt_to_usmc()`](https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/fn.sealed_sender_encrypt_from_usmc.html)
- [`sealed_sender_multi_recipient_encrypt()`](https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/fn.sealed_sender_multi_recipient_encrypt.html)

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +803,58 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+///
+/// #### Wire Format
+/// The output of this method is encoded as a protobuf with an additional [u8] prepended to the
+/// output indicating that this message was sent using the single-recipient KEM scheme.
```
I think it's just "indicating the version of Sealed Sender in use".

cosmicexplorer replied:
Thanks! Changed.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1194,14 +1491,18 @@ impl SealedSenderDecryptionResult {
 
+/// Decrypt a sealed sender message's sender information, then decrypt the underlying message.
+///
+/// Decrypt the sealed sender information from `ciphertext` using [sealed_sender_decrypt_to_usmc],
+/// drop the message if it was sent to the local user at `local_address`, then decrypt the
+/// underlying message according to [UnidentifiedSenderMessageContent::msg_type].
```
I think this is at once more detail and less detail than a caller of this function would want. The self-send failure path is one of multiple kinds of errors that can happen, and the introduction of UnidentifiedSenderMessageContent doesn't really make sense without clicking through to `sealed_sender_decrypt_to_usmc`. (The one really important thing this method does that `sealed_sender_decrypt_to_usmc` does not is validate the embedded sender certificate; without that, you might have a forged certificate and therefore a forged sender.)

In another PR, it would probably be reasonable to take `local_address` as `Option<&ProtocolAddress>`, to skip the local address checking or reuse an existing object. That's too big a change for here, though. Stay on target!

cosmicexplorer replied:
> (The one really important thing this method does that `sealed_sender_decrypt_to_usmc` does not is validate the embedded sender certificate; without that, you might have a forged certificate and therefore a forged sender.)

Thanks! I have removed this information entirely and focused the docstring purely on the meaning of the sender certificate validation (which is much more brief).

> In another PR, it would probably be reasonable to take `local_address` as `Option<&ProtocolAddress>`, to skip the local address checking or reuse an existing object. That's too big a change for here, though. Stay on target!

Thanks! I have reverted that change.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +803,58 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+/// The KEM scheme implemented by this method is described in [this Signal blog post]. The
+/// high-level steps of this process are listed below:
+/// 1. Generate a random key pair.
+/// 2. Derive a symmetric chain key, cipher key, and MAC key from a salt composed of the recipient's
+///    public key and the sender's public and private key pair.
```
HKDF takes multiple inputs; they're not all salt. I'd just drop "a salt composed of".

cosmicexplorer replied:
Thanks, done!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -853,6 +990,21 @@ mod sealed_sender_v2 {
+    ///
+    /// *See more information on [authenticated encryption].*
+    ///
+    /// [authentication tag]: https://crypto.stackexchange.com/a/67108
+    /// [authenticated encryption]: https://en.wikipedia.org/wiki/Authenticated_encryption
```
I think you could just link to the Wikipedia article to explain "authentication tag" instead of picking a specific StackExchange post. Feels weird to see that in a codebase to explain a generic crypto term.

cosmicexplorer replied:
Yes, this makes sense. I think it's because wikipedia doesn't immediately redirect "authentication tag" to "message authentication code" and I didn't do enough looking around. I agree it feels safer to link to wikipedia in general.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1031,133 @@ mod sealed_sender_v2 {
+    }
 }
 
+/// This method implements a single-key multi-recipient [KEM], defined in Manuel Barbosa's
```
"as defined", I think, since we're using a particular set of parameters for Barbosa's paper.

cosmicexplorer replied:
Thanks! Changed!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1031,133 @@ mod sealed_sender_v2 {
+/// ### Contrast with Single-Recipient KEM
+/// The KEM scheme implemented by this method uses the "Generic Construction" in `4.1` of that
+/// paper, instantiated with [ElGamal encryption]. Unlike other message types sent over the wire,
+/// the encoded message returned by this method does *not* correspond to a specific protobuf schema
+/// due to the complex web of cryptographic data dependencies in different segments of the resulting
+/// encoded message. See the documentation of our [single-recipient KEM] for a discussion of the
+/// design tradeoffs of using this particular multi-recipient KEM.
```
Eh. It could have been protobuf, but that's just wasting space in a format with so many fixed-width fields. We went back and forth but decided that the only thing protobuf would buy us is extensibility, and we could add another version instead.

I wrote up this document internally on the format; maybe some form of it belongs here or in the repository somewhere as well: 
[multi_recipient_sealed_sender.md](https://github.com/signalapp/libsignal-client/files/7255544/multi_recipient_sealed_sender.md).

cosmicexplorer replied:
The rustdoc book mentions it does not have support for bare markdown files, which would certainly be convenient here. I think this looks brief and appropriate enough for the "Wire Format" section of the `sealed_sender_multi_recipient_encrypt()` docstring, so I'll try adapting it there.

cosmicexplorer replied:
Hmmmm, rust-lang/rust#76709 certainly seems to imply it is possible to use rustdoc against markdown files directly. I will try to apply that feature.

cosmicexplorer replied:
Ok, so it's possible to use *rustdoc* against markdown files directly as a command-line argument, but it seems that isn't integrated into `cargo doc` (either providing files as CLI arguments, or using raw markdown files). In 94d7ca0 I took the liberty of merging this doc into the "Wire Format" section of `sealed_sender_multi_recipient_encrypt()`, and linking into the resulting "The version byte" section from the "Wire Format" section of `sealed_sender_encrypt_from_usmc()`.

In these changes I have also referred to "Sealed Sender v1" and "Sealed Sender v2" instead of "the single-/multi-recipient KEM". See the "Wire Format" section for Sealed Sender v2 rendered at https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/fn.sealed_sender_multi_recipient_encrypt.html#wire-format.

jrose-signal replied:
I wasn't necessarily thinking "in the repository" had to mean "included in the rustdoc", but it seems to work, especially since functions get their own pages. (Thanks for rendering the docs!) You *can* include a Markdown file using `#[doc = include_str!("foo.md")]` (if we're on a new enough Rust, h/t [the Rustdoc book](https://doc.rust-lang.org/rustdoc/the-doc-attribute.html)), but I think this is better anyway.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1031,133 @@ mod sealed_sender_v2 {
+/// 4. Add an authentication tag with for (3) using a secret derived from the sender's private
+///    identity key and the recipient's public identity key.
+/// 5. Use the keypair from (2) to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent] via [AEAD encryption].
```
Let's add that this only has to be done once per recipient, unlike (3) and (4).

cosmicexplorer replied:
Thanks! Done.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +803,58 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+/// [KEM]: https://en.wikipedia.org/wiki/Key_encapsulation
+/// [this Signal blog post]: https://signal.org/blog/sealed-sender/
+///
+/// ### Contrast with Multi-Recipient KEM
```
Extreme nitpick: the rustdoc book seems to suggest starting at `#` for section headers, even if it's pretty strong.

cosmicexplorer replied:
Done! I believe it is possible to edit the CSS used in a generated docsite if we find it too aggressive (it seems mostly fine to look at by default though), although I haven't done so myself yet.

jrose-signal reviewed (commented):
Pile of comments but I appreciate this work! I think you've done well separating the public and private parts.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -757,31 +757,43 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
+///    a secret derived from the sender's private identity key and the recipient's public
+///    identity key.
+/// 5. Use the keypair from (2) to symmetrically encrypt the underlying
 ///    [UnidentifiedSenderMessageContent].
```
Yeah, I think I got to that realization as well. I tried deleting this numbered list section and I *think* I still prefer having it, but we can probably afford to be more concise since the pseudocode + paper link is actually extremely useful.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +595,194 @@ impl UnidentifiedSenderMessage {
+                    their_public.serialize(),
+                ],
             }
+            .concat();
```
Done!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +595,194 @@ impl UnidentifiedSenderMessage {
+                .private_key()
+                .calculate_agreement(their_key.public_key())?;
+            // 96 bytes are derived, but the first 32 are discarded/unused.
+            // TODO: why?
```
Thanks! Left a quick comment to that regard, let me know if that's extraneous info.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +595,194 @@ impl UnidentifiedSenderMessage {
+    fn test_agreement_and_authentication() -> Result<()> {
+        // The sender and recipient each have a long-term identity key pair.
+        let sender_keypair = KeyPair::generate(&mut rand::thread_rng());
+        let sender_identity = IdentityKeyPair::try_from(sender_keypair).unwrap();
```
Thanks! Fixed.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +803,58 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+///    [crate::process_prekey_bundle] after an initial [crate::PreKeySignalMessage] is recieved.
+/// 2. it ferries a lot of additional information in its encoding which makes the resulting message
+///    bulkier than the message produced by this method.
+/// 3. it does not currently correspond to a specific protobuf schema.
```
Thanks! Done.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -720,29 +871,28 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
-        eph_keys.cipher_key()?,
-        eph_keys.mac_key()?,
+        eph_keys.cipher_key.as_ref(),
+        eph_keys.mac_key.as_ref(),
```
Yes! Fixed.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -821,6 +944,15 @@ mod sealed_sender_v2 {
         }
     }
 
+    /// Encrypt a slice of random bytes `input` using a shared secret derived from the ephemeral
```
> "Encrypt or decrypt" maybe, since it's XOR?

Yes! Done.

> maybe `input` should be one too

Yes! I ended up putting a ton of other changes relating to fixed-size arrays and removal of unnecessary `Result`s into my `docs-3` branch which is intended to follow up on #287; see https://github.com/cosmicexplorer/libsignal-client/compare/docs-2...cosmicexplorer:docs-3?expand=1. I can definitely make that change here though.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1006,7 +1278,18 @@ pub async fn sealed_sender_multi_recipient_encrypt<R: Rng + CryptoRng>(
+/// individual encrypted [UnidentifiedSenderMessageContent]s. **Note: this method is only used in
+/// testing.**
+///
+/// This method is used for testing in the context of the Signal app, to allow processing a bulk v2
```
This was redundant, as I hadn't clarified yet how Signal processes the contents of this message, and I have removed it. It was intended to be a vague placeholder until I had figured that out.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1006,7 +1278,18 @@ pub async fn sealed_sender_multi_recipient_encrypt<R: Rng + CryptoRng>(
+/// sealed-sender message with one or more recipients directly by
+/// [sealed_sender_decrypt_to_usmc]. For the Signal app, this process of splitting out a v2
+/// sealed-sender message into individual messages and correctly routing the result to recipients is
+/// performed implicitly by the Signal server, and not in this codebase.
```
Thanks! Removed.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1006,7 +1278,18 @@ pub async fn sealed_sender_multi_recipient_encrypt<R: Rng + CryptoRng>(
+/// performed implicitly by the Signal server, and not in this codebase.
+///
+/// Note: It does not appear to be possible to expose a method declared with `#[cfg(test)]` to
+/// another module, even if that module is itself being compiled and executed as a test.
```
You're right! I was more looking to make sure nobody tried to add `#[cfg(test)]` later and wasted time trying to figure out why it wouldn't work. I first moved it into a non-docstring comment, but that seemed to make it more confusing, so I've instead modified the docstring in light of your explanation above to clarify that this method *strips metadata needed for routing messages* and that's why the Signal server processes it differently.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1060,6 +1343,11 @@ pub fn sealed_sender_multi_recipient_fan_out(data: &[u8]) -> Result<Vec<Vec<u8>>
+///
+/// This method produces an [UnidentifiedSenderMessageContent] with
+/// the plaintext information about the sender's identity, which is then used to finally decrypt the
+/// underlying encrypted message in [sealed_sender_decrypt].
```
I have clarified that this method extracts the sender's identity from the underlying message and removed the extraneous mention of `[UnidentifiedSenderMessageContent]`, since that is in the return type of the method.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1074,31 +1362,30 @@ pub async fn sealed_sender_decrypt_to_usmc(
             )?;
 
-            let static_key = PublicKey::try_from(&message_key_bytes[..])?;
+            let static_key = IdentityKey::from(PublicKey::try_from(&message_key_bytes[..])?);
```
Ok! I assume that's because it could be an ephemeral key which isn't actually tied to anyone's identity? Should the `IdentityKeyPair` also be changed?

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +595,194 @@ impl UnidentifiedSenderMessage {
+        /// recipient's identities, as well as [EphemeralKeys::chain_key].
+        pub(super) fn calculate(
+            our_keys: &IdentityKeyPair,
+            their_key: &IdentityKey,
```
Changed--see comment earlier https://github.com/signalapp/libsignal-client/pull/366#discussion_r719810457.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -910,23 +910,28 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
-    const LABEL_K: &[u8] = b"Sealed Sender v2: K";
-    const LABEL_DH: &[u8] = b"Sealed Sender v2: DH";
-    const LABEL_DH_S: &[u8] = b"Sealed Sender v2: DH-sender";
+    /// Static byte strings used as part of a MAC in [HKDF::derive_secrets].
```
> It's more about getting different values out for the same key material.

That makes perfect sense! The term MAC was definitely used in a pretty vague sense which I realize is probably not a good idea for specific security terms :)

> Anyway, I think this is probably over-organizing. We don't have a policy to not use regular comments,

I was concerned about this! I have removed the `mod` wrapping and converted it into a regular comment.

> We don't have a policy to not use regular comments,

Thanks! Will keep this in mind.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+            Direction::Sending,
+            &m,
+        )?
+        .as_ref()
```
I don't think you need this `as_ref()` at all; [Deref](https://doc.rust-lang.org/std/ops/trait.Deref.html#more-on-deref-coercion) should take care of it.

cosmicexplorer replied:
Looks to be so!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+        let recipient_identity = IdentityKeyPair::generate(&mut rand::thread_rng());
+
+        // Generate random bytes used for our multi-recipient encoding scheme.
+        let m: [u8; MESSAGE_KEY_LEN] = (&mut rand::thread_rng()).gen();
```
`rand::thread_rng().gen()` should work, `&` or `&mut` are implicitly added for method calls even for return values.

cosmicexplorer replied:
Thanks for the explanation! That's correct!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+            Direction::Receiving,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_m.as_ref(), m.as_ref());
```
`&`

cosmicexplorer replied:
Fixed!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+            &ephemeral_public_key,
+            &sender_c_0,
+        )?;
+        assert_eq!(recv_at_0.as_ref(), sender_at_0.as_ref());
```
`&`

cosmicexplorer replied:
Fixed!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// ["Randomness Reuse: Extensions and Improvements"]: https://haslab.uminho.pt/mbb/files/reuse.pdf
+///
+/// # Contrast with Sealed Sender v1
+/// The KEM scheme implemented by this method uses the "Generic Construction" in `4.1` of [that
+/// paper]["Randomness Reuse: Extensions and Improvements"], instantiated with [ElGamal
```
Nitpick: since this is in its own section now, let's use "Barbosa's paper" instead of "that paper".

cosmicexplorer replied:
Thanks!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// 1. it requires a [SessionRecord] to exist already for the recipient, i.e. that a Double Ratchet
+///    message chain has previously been established in the [SessionStore] via
+///    [crate::process_prekey_bundle] after an initial [crate::PreKeySignalMessage] is recieved.
+/// 2. it ferries a lot of additional information in its encoding which makes the resulting message
+///    bulkier than the message produced by this method.
```
"this method" is left over from v1, but in addition I'll add a second sentence of my own: "For sending, this will generally still be more compact than sending the same message N times, but on the receiver side the message is slightly larger."

cosmicexplorer replied:
Yes, that's much better!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+///    [crate::process_prekey_bundle] after an initial [crate::PreKeySignalMessage] is recieved.
+/// 2. it ferries a lot of additional information in its encoding which makes the resulting message
+///    bulkier than the message produced by this method.
+/// 3. unlike other message types sent over the wire, the encoded message returned by this method
+///    does *not* correspond to a specific protobuf schema, in order to avoid inefficiencies
+///    produced by protobuf's packing (see **[Wire Format]**).
```
I'd just say "does not use protobuf" rather than "does not correspond to a specific protobuf schema". It's not protobuf-compatible at all.

cosmicexplorer replied:
That makes sense!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+///    produced by protobuf's packing (see **[Wire Format]**).
+///
+/// [ElGamal encryption]: https://en.wikipedia.org/wiki/ElGamal_encryption
+/// [Sealed Sender v1]: sealed_sender_encrypt_from_usmc#contrast-with-sealed-sender-v2
```
I'd drop the anchor here and just link to the function; it's the first section and a reader would probably want to know what's being contrasted with before they look at the reference back to SSv2.

cosmicexplorer replied:
That makes sense, especially given that we're just calling it "Sealed Sender v1" now!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// [Sealed Sender v1]: sealed_sender_encrypt_from_usmc#contrast-with-sealed-sender-v2
+/// [Wire Format]: #wire-format
+///
+/// # Overview
```
"Algorithmic Overview" again

cosmicexplorer replied:
Have also converted this to "High-level algorithmic overview" -- let me know if that's reasonable or too much!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// 5. Generate a symmetric key from (1) and use it to symmetrically encrypt the underlying
+///    [UnidentifiedSenderMessageContent] via [AEAD encryption]. *This step is only performed once
+///    per message, regardless of the number of recipients.*
+/// 6. Send the public ephemeral key (2), the encrypted random bytes (3), and the authentication tag
+///    (4) to the server, along with the encrypted message (5).
```
"authentication tags" should be plural, since there's one per recipient. (I guess we can't double-pluralize "random bytes" though.)

cosmicexplorer replied:
I've used the phrasing "the sequence of encrypted random bytes and authentication tags" to clarify this concisely here!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// which may make it comparatively unwieldy for certain scenarios:
+/// 1. it requires a [SessionRecord] to exist already for the recipient, i.e. that a Double Ratchet
+///    message chain has previously been established in the [SessionStore] via
+///    [crate::process_prekey_bundle] after an initial [crate::PreKeySignalMessage] is recieved.
```
Everything's still flat for now, let's ditch the `crate::`.

cosmicexplorer replied:
Done!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
````diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+///     return message
+///```
+///
+/// # Wire Format
````
In v1 you had this as a subsection of the algorithmic overview, and I think that's reasonable. Either way they should be consistent, though.

cosmicexplorer replied:
I have made both of these sections into `h1`s, let me know if that looks reasonable!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+///
+/// ## Routing messages to recipients
+///
+/// The Signal backend server uses the initial version byte from the [sent message][sending] to
+/// determine whether the message was sent using Sealed Sender v2. If so, the server will then split
+/// up the set of messages and securely route each individual [received message][receiving] to its
+/// intended recipient.
```
Honestly they're totally separate entry points, because a v1 message doesn't include its recipient in the encoding. I'd skip ahead to "the server will split up…".

cosmicexplorer replied:
Done!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// For testing purposes, [sealed_sender_multi_recipient_fan_out] can be used to convert such a bulk
+/// message produced by Sealed Sender v2 into a sequence of [received messages][receiving]; however,
+/// in doing so it will drop all of the metadata necessary to identify the message's
+/// intended recipient.
```
"recipients"

cosmicexplorer replied:
Done!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +798,55 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+/// The *single-recipient* KEM scheme implemented by this method partially derives the encryption
+/// key from the recipient's identity key, which would then require re-encrypting the same message
+/// multiple times to send to multiple recipients. In contrast,
+/// [Sealed Sender v2][sealed_sender_multi_recipient_encrypt] uses a *multi-recipient* KEM scheme
```
I believe this link should be using parens, not brackets, since `sealed_sender_multi_recipient_encrypt` is a destination. (Maybe it's a rustdoc extension? But still.)

cosmicexplorer replied:
Ah yes! It is a rustdoc extension, but it appears to work great with parens!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +798,55 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+/// The output of this method is encoded as an `UnidentifiedSenderMessage.Message` from
+/// `sealed_sender.proto`, prepended with an additional byte to indicate the version of Sealed
+/// Sender in use (see [further documentation on the version
+/// byte][sealed_sender_multi_recipient_encrypt#the-version-byte]). In all cases,
```
Here too.

cosmicexplorer replied:
Done!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -853,6 +978,21 @@ mod sealed_sender_v2 {
     }
 
+    /// Compute an [authentication tag] for the encrypted result `encrypted_message_key` produced by
+    /// [apply_agreement_xor] using a secret derived from the sender's private identity key
```
Rust convention seems to be to use code voice even within links (and rustdoc accepts it with no additional annotation): ``[`apply_agreement_xor`]``. There are a *bunch* of places to change this, sorry!

cosmicexplorer replied:
Not a problem, emacs makes it easy!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// [sending]: #sent-messages
+/// [receiving]: #received-messages
+///
+/// ## The version byte
```
This belongs within the Wire Format section…

cosmicexplorer replied:
Yes! Have made the Wire Format section into a top-level heading as was intended.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// of `0x22`. A hypothetical version byte `0x34` would indicate a message encoded
+/// as Sealed Sender v4, but decodable by any client that supports Sealed Sender v3.
+///
+/// ## Routing messages to recipients
```
…but this doesn't…

cosmicexplorer replied:
I have moved it into its own top-level heading above Wire Format!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// in doing so it will drop all of the metadata necessary to identify the message's
+/// intended recipient.
+///
+/// ## Received messages
```
…and this does again.

cosmicexplorer replied:
This is now a subsection of Wire Format after moving Routing messages to recipients into its own top-level heading!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
````diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+/// ```
+///
+/// The varint encoding used is the same as [protobuf's][varint]. Values are unsigned. UUIDs are
+/// encoded per [RFC 4122], with the first eight bytes considered "most significant" [^1].
````
Typographical nitpick: footnotes after a sentence should follow the period, not precede it.

cosmicexplorer replied:
I did not realize that--it's very reasonable to me from Wikipedia formatting! Fixed!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -879,8 +1019,214 @@ mod sealed_sender_v2 {
+///
+/// The varint encoding used is the same as [protobuf's][varint]. Values are unsigned. UUIDs are
+/// encoded per [RFC 4122], with the first eight bytes considered "most significant" [^1].
+/// Fixed-width integers are in network byte order (big-endian).
```
Thought of something I forgot to put in the original doc: "Fixed-width integers are unaligned and in network byte order".

cosmicexplorer replied:
Made this change!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1001,12 +1353,22 @@ pub async fn sealed_sender_multi_recipient_encrypt<R: Rng + CryptoRng>(
+/// testing.**
+///
+/// This method strips recipients' metadata and splits a bulk v2 sealed-sender message into byte
+/// strings which can be processed by [sealed_sender_decrypt_to_usmc]. This is used in our
+/// Java/Swift/TypeScript tests. For the Signal app, this process of splitting out a v2
```
I'd strike "this is used in our tests". It's used in our Rust tests too! It's free for clients to use! The note above gets the point across already.

cosmicexplorer replied:
This makes sense! The point about stripping the metadata needed to route the message was I think all I wanted to make clear in my original attempt at this documentation change, so I have removed this sentence!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1060,6 +1422,10 @@ pub fn sealed_sender_multi_recipient_fan_out(data: &[u8]) -> Result<Vec<Vec<u8>>
     Ok(messages)
 }
 
+/// Decrypt the sender's identity from a sealed-sender message in either the v1 or v2 format.
+///
+/// [sealed_sender_decrypt] consumes the output of this method to validate the sender's identity
+/// before decrypting the underlying message.
```
"Decrypts the payload of a sealed-sender message in either the v1 or v2 format". We don't really have language around UnidentifiedSenderMessageContents, but it's not just the sender's identity. It has the real message in there too.

cosmicexplorer replied:
This makes sense!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1074,31 +1440,30 @@ pub async fn sealed_sender_decrypt_to_usmc(
                 &encrypted_static,
-                eph_keys.cipher_key()?,
-                eph_keys.mac_key()?,
+                eph_keys.cipher_key.as_ref(),
+                eph_keys.mac_key.as_ref(),
```
`&`

cosmicexplorer replied:
Done! Got to this before seeing this comment :)

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1074,31 +1440,30 @@ pub async fn sealed_sender_decrypt_to_usmc(
                 &encrypted_message,
-                static_keys.cipher_key()?,
-                static_keys.mac_key()?,
+                static_keys.cipher_key.as_ref(),
+                static_keys.mac_key.as_ref(),
```
`&`

cosmicexplorer replied:
Done! Got to this before seeing this comment too!

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1117,6 +1482,15 @@ pub async fn sealed_sender_decrypt_to_usmc(
             encrypted_message,
         } => {
+            let encrypted_message_key: [u8; 32] = encrypted_message_key
+                .as_ref()
```
(unneeded again)

cosmicexplorer replied:
Hmmm, I'm getting this when I try to remove it:
```rust
error[E0277]: the trait bound `[u8; 32]: From<Box<[u8]>>` is not satisfied
    --> rust/protocol/src/sealed_sender.rs:1483:39
     |
1483 |                 encrypted_message_key.try_into().map_err(|_| {
     |                                       ^^^^^^^^ the trait `From<Box<[u8]>>` is not implemented for `[u8; 32]`
```

jrose-signal replied:
Oh, hm, fair. Okay, `as_ref()` is reasonable here.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1117,6 +1482,15 @@ pub async fn sealed_sender_decrypt_to_usmc(
+                .as_ref()
+                .try_into()
+                .map_err(|_: TryFromSliceError| {
+                    SignalProtocolError::BadKeyLength(
+                        curve::KeyType::Djb,
+                        encrypted_message_key.len(),
```
Hmm. The message key isn't a curve key. Let's just go with InvalidSealedSenderMessage for now; the errors need attention separately anyway, and in practice this should never happen.

cosmicexplorer replied:
That makes sense! I've done this with a `format!(...)` string which prints the detected size of that slice although you've mentioned it should never happen.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1194,6 +1569,12 @@ impl SealedSenderDecryptionResult {
+/// Decrypt a Sealed Sender message `ciphertext` in either the v1 or v2 format, and validate its
+/// sender certificate.
+///
+/// This method calls [sealed_sender_decrypt_to_usmc] to extract the sender information, including
+/// the embedded [SenderCertificate]. The sender certificate is then validated against the
+/// `trust_root` provided by the backend server to ensure that the sender's identity was not forged.
```
I think it's worth adding something like "For more details on the sealed sender formats, see [`sealed_sender_encrypt_from_usmc`] and [`sealed_sender_multi_recipient_encrypt`]."

jrose-signal replied:
`trust_root` isn't actually provided by the server; it's baked into the apps. The root signs the server certificate, which signs the sender certificate.

cosmicexplorer replied:
Thank you for this very very helpful clarification!! I have added both of these comments here.

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -1194,6 +1569,12 @@ impl SealedSenderDecryptionResult {
     }
 }
 
+/// Decrypt a Sealed Sender message `ciphertext` in either the v1 or v2 format, and validate its
+/// sender certificate.
```
", validate its sender certificate, and then decrypt the inner message payload". That's the most important part! :-)

cosmicexplorer replied:
Thanks for clarifying this part!! Definitely a necessary part of the method description :)

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
 }
 
+/// Encrypt the plaintext message `ptext`, generate an [UnidentifiedSenderMessageContent], then pass
+/// the result to [sealed_sender_encrypt_from_usmc].
```
It's pretty clear that `sealed_sender_encrypt_from_usmc` is the place to go from the summary, but I think it could still stand a body line like "This is a simple way to encrypt a message in a 1:1 session using [Sealed Sender v1]." It's just a shortcut, albeit one that made sense as presented API before SenderKey came around.

cosmicexplorer replied:
Thanks for this clarification! I've made this change! I will probably try to parse/document the `SenderKey` mechanism next :)))

jrose-signal reviewed (commented):
Okay, this one's nearly all copyediting-type comments—all the structure looks good to me now. I'm sorry there are so many! But it's definitely converging. And again I appreciate you having generated the docs; that makes it much easier to read and review.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+            let their_pub_key = their_public.serialize();
+            let ephemeral_salt = match direction {
+                Direction::Sending => [SALT_PREFIX, their_pub_key.as_ref(), our_pub_key.as_ref()],
+                Direction::Receiving => [SALT_PREFIX, our_pub_key.as_ref(), their_pub_key.as_ref()],
```
Done! Thanks!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+            let shared_secret = our_keys.private_key().calculate_agreement(their_key)?;
+            // 96 bytes are derived, but the first 32 are discarded/unused. This is intended to
+            // mirror the way the EphemeralKeys are derived, even though StaticKeys does not end up
+            // requiring a third "chain key".
```
:D

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+        let sender_static_key_ctext = crypto::aes256_ctr_hmacsha256_encrypt(
+            &sender_identity.public_key().serialize(),
+            sender_eph_keys.cipher_key.as_ref(),
+            sender_eph_keys.mac_key.as_ref(),
```
Done! Thanks!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+        let sender_message_data = crypto::aes256_ctr_hmacsha256_encrypt(
+            sender_message_contents,
+            sender_static_keys.cipher_key.as_ref(),
+            sender_static_keys.mac_key.as_ref(),
```
Fixed!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+        let recipient_message_key_bytes = crypto::aes256_ctr_hmacsha256_decrypt(
+            &sender_static_key_ctext,
+            recipient_eph_keys.cipher_key.as_ref(),
+            recipient_eph_keys.mac_key.as_ref(),
```
Done! Thanks!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+        let recipient_message_contents = crypto::aes256_ctr_hmacsha256_decrypt(
+            &sender_message_data,
+            recipient_static_keys.cipher_key.as_ref(),
+            recipient_static_keys.mac_key.as_ref(),
```
Done! Thanks!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +798,55 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
 }
 
+/// This method implements the single-key single-recipient [KEM] described in [this Signal blog
+/// post], a.k.a Sealed Sender v1.
```
Done! Thanks!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +798,55 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+/// Sender in use (see [further documentation on the version
+/// byte][sealed_sender_multi_recipient_encrypt#the-version-byte]). In all cases,
+/// [sealed_sender_decrypt] is used in the client to decrypt the Sealed Sender message, regardless
+/// of version.
```
I wasn't sure how to fix that--thanks! Agree that it is relevant info that should be higher up in the docstring.

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -758,38 +900,9 @@ pub async fn sealed_sender_encrypt_from_usmc<R: Rng + CryptoRng>(
-
     use super::*;
 
+    // Static byte strings used as part of a MAC in [HKDF::derive_secrets].
```
Thanks! That's much better!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -853,6 +978,21 @@ mod sealed_sender_v2 {
 
+    /// Compute an [authentication tag] for the encrypted result `encrypted_message_key` produced by
+    /// [apply_agreement_xor] using a secret derived from the sender's private identity key
+    /// `our_keys` and the recipient's public identity key `their_key`.
```
Fixed!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -821,14 +938,22 @@ mod sealed_sender_v2 {
 
+    /// Encrypt or decrypt a slice of random bytes `input` using a shared secret derived from the
+    /// ephemeral [DerivedKeys::e] `our_keys` and the message recipient's public identity key
+    /// `their_key`.
```
I have drastically reduced the complexity of this description, just clarifying how calling with `Direction::Receiving` inverts the encryption performed by calling with `Direction::Sending`!

Note: I have also taken the liberty of making `apply_agreement_xor` return `[u8; MESSAGE_KEY_LEN]` and `compute_authentication_tag` return `[u8; AUTH_TAG_LEN]`!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -704,6 +798,55 @@ pub async fn sealed_sender_encrypt<R: Rng + CryptoRng>(
+/// [Sealed Sender v2][sealed_sender_multi_recipient_encrypt] uses a *multi-recipient* KEM scheme
+/// which avoids this repeated work, but makes a few additional design tradeoffs.
+///
+/// # Overview
```
Changed to "High-level algorithmic overview" to make it exceedingly clear this is not intended as a formal description!

cosmicexplorer commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+            recipient_eph_keys.cipher_key.as_ref(),
+            recipient_eph_keys.mac_key.as_ref(),
+        )?;
+        let sender_public_key = PublicKey::try_from(recipient_message_key_bytes.as_ref())?;
```
Hmmm, I get this error when I try to use a bare `&`:
```rust
error[E0277]: the trait bound `curve::PublicKey: From<Vec<u8>>` is not satisfied
   --> rust/protocol/src/sealed_sender.rs:757:44
    |
757 |         let sender_public_key: PublicKey = PublicKey::try_from(recipient_message_key_bytes)?;
    |                                            ^^^^^^^^^^^^^^^^^^^ the trait `From<Vec<u8>>` is not implemented for `curve::PublicKey`
```

jrose-signal replied:
Ah, right. You could spell it `&foo[..]` instead of `foo.as_ref()` but it's not the obvious change the others were. (I wish there were a Clippy lint for this, so I wasn't guessing from looking at the code.)

cosmicexplorer replied:
I will note that! I have always wanted to make a clippy lint :)))

cosmicexplorer commented:
Ok, I've addressed most of your comments and left 9 of them unresolved for further review! As before, the compiled `cargo doc` output has been pushed to https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/, and the focus of this PR is on these two:
- Sealed Sender v1: [`sealed_sender_encrypt_from_usmc()`](https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/fn.sealed_sender_encrypt_from_usmc.html)
- Sealed Sender v2: [`sealed_sender_multi_recipient_encrypt()`](https://cosmicexplorer.github.io/libsignal-client/target/doc/libsignal_protocol/fn.sealed_sender_multi_recipient_encrypt.html)

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -877,10 +1015,219 @@ mod sealed_sender_v2 {
+        boxed_result
+            .as_ref()
+            .try_into()
+            .map_err(|_| SignalProtocolError::BadKeyLength(curve::KeyType::Djb, boxed_result.len()))
```
Same thing about not being a Curve25519 key, but also I think it's okay to `expect()` this one because we just generated it from HKDF; that is, it's not data passed in from outside. (Really I need to change HKDF's interface to generate into an output slice instead of returning a box, and then all these conversions go away. I have a half-written patch for that already so you don't need to do it.)

cosmicexplorer replied:
Done! Changes are in cf048dd.

> I have a half-written patch for that already so you don't need to do it.

Yay!! I remember that part was a lot of work when I tried to tackle it :)

jrose-signal commented on rust/protocol/src/sealed_sender.rs:
```diff
@@ -821,14 +943,18 @@ mod sealed_sender_v2 {
         }
     }
 
+    /// Encrypt or decrypt a slice of random bytes `input` using a shared secret derived from
+    /// `our_keys` and `their_key`.
+    ///
+    /// The output of this method when called with [`Direction::Sending`] can be inverted to produce
+    /// the original `input` bytes if called with [`Direction::Receiving`] with `our_keys` and
+    /// `their_key` swapped.
```
I like these simple descriptions a lot better, thank you.

cosmicexplorer replied:
:DDDD

jrose-signal reviewed (commented):
Answered your open questions (mostly in agreement) and I think these'll be my last comments!

cosmicexplorer commented:
Ok, cf048dd integrates your notes on `&...[..]` and `.expect("...")` usage! It does not change any docstrings, so I have not pushed any docs changes since this comment https://github.com/signalapp/libsignal-client/pull/366#issuecomment-933040985.

I'll also probably do a once-over again and see if there's anything else I think could be improved. I think the result is DRASTICALLY better than my initial attempt here and I learned quite a lot from this PR :DDDD. Thanks again for that.

cosmicexplorer commented:
After reviewing the docs again, I *think* I can sign off on this so merge it whenever! Thanks again for your time making this amazing, it helped me a lot and I hope many others too :)))
